README.md
# path_set [![Flattr this][flatter_png]][flatter]

Modify system and user environment variables.

## Usage

````
path_set [options] path [path2 ...]

Options:
-v VarName Append path(s) to environmental variable VarName (default is Path)
-s         Modify variable in the system environment (default)
-u         Modify variable in the current user's environment
-d         Modify variable in the default user environment
-e char    Change % substitution character to 'char' (default is `)
-n         Don't add path if the variable already contains it
-p         Prepend path(s) to variable
-r         Replace the value (default is to append)
-x         Use REG_EXPAND_SZ type (default if ` char is used)
-z         Use REG_SZ type (default)
-c         Show version and copyright statement
-?         Show this help screen
````

## Examples

````batch
  rem append c:\bin to the system Path variable:
  path_set -n c:\bin

  rem prepend c:\usr\local\bin and c:\bin to the system Path:
  path_set -n -p c:\usr\local\bin c:\bin

  rem Add "C:\Program Files\Java\bin" to the system Path:
  path_set -n "C:\Program Files\Java\bin"
````

or, even better:

````batch
  rem add or replace JAVA_HOME with %ProgramFiles%\Java
  path_set -n -r -x -v JAVA_HOME `ProgramFiles`\Java

  rem Append %JAVA_HOME%\bin to the system Path:
  path_set -n `JAVA_HOME`\bin
````

## Verify a Release

To verify a release, download the .zip, .sha256, and .asc files for the release 
(replacing path_set-1.5-win32.zip with the release you are verifying):

````
$ wget https://github.com/rasa/path_set/releases/download/v1.5/path_set-1.5-win32.zip{,.sha256,.asc}
````

Next, check that sha256sum reports "OK":
````
$ sha256sum -c path_set-1.5-win32.zip.sha256
path_set-1.5-win32.zip: OK
````

Lastly, check that GPG reports "Good signature":

````
$ gpg --keyserver hkps.pool.sks-keyservers.net --recv-key 0x105a5225b6ab4b22
$ gpg --verify path_set-1.5-win32.zip.asc path_set-1.5-win32.zip
gpg:                using RSA key 0xFF914F74B4BB6EF3
gpg: Good signature from "Ross Smith II <ross@smithii.com>" [ultimate]
...
````

## Contributing

To contribute to this project, please see [CONTRIBUTING.md](CONTRIBUTING.md).

## Bugs

To view existing bugs, or report a new bug, please see [issues](../../issues).

## Changelog

To view the version history for this project, please see [CHANGELOG.md](CHANGELOG.md).

## License

This project is [MIT licensed](LICENSE).

## Contact

This project was created and is maintained by [Ross Smith II][] [![endorse][endorse_png]][endorse]

Feedback, suggestions, and enhancements are welcome.

[Ross Smith II]: mailto:ross@smithii.com "ross@smithii.com"
[flatter]: https://flattr.com/submit/auto?user_id=rasa&url=https%3A%2F%2Fgithub.com%2Frasa%2Fpath_set
[flatter_png]: http://button.flattr.com/flattr-badge-large.png "Flattr this"
[endorse]: https://coderwall.com/rasa
[endorse_png]: https://api.coderwall.com/rasa/endorsecount.png "endorse"



CHANGELOG.md
## 1.7 (unreleased)

  * Update Makefile
  
## 1.6 (28-Mar-2015)

  * Fix segfault due to linker issue (#1)

## 1.5 (27-Mar-2015)

  * Executables signed with StartSSL code signing certificate
  * Release zip files are gpg signed and sha256 hashed
  * Standardize build scripts
  * Update documentation

## 1.4 (22-Feb-2015)

  * Improve command line option processing
  * Updated to Visual Studio 2013
  * Switch to MIT license

## 1.3 (12-Jul-2006)

  * Changed ~ character to `, as ~ is common in 8.3 paths
  * Standardized version information across all programs

## 1.2 (08-Apr-2006)

  * Added .rc and .ico files
  * Improved build scripts
  * Updated documentation

## 1.1 (17-Feb-2006)

  * Renamed from AddPath to path_set

## 1.0 (06-Jan-2006)

  * Initial release


UpdatePathVariable.cpp
// Copyright (c) 2005-2015 Ross Smith II. See Mit LICENSE in /LICENSE

/*
todo WM_SETTINGCHANGE isn't working right, research
*/

#ifdef _MSC_VER
#pragma warning(disable:4996)
#endif

#define _WIN32_WINNT 0x501

#ifndef WIN32_LEAN_AND_MEAN
# define WIN32_LEAN_AND_MEAN 1
#endif

#include <windows.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "UpdatePathVariable.h"
#include "debug.h"
#include "SystemMessage.h"

#define MAX_VALUE_LEN  32767

#define ENVREGKEY       "System\\CurrentControlSet\\Control\\Session Manager\\Environment"
#define USERENVREGKEY   "Environment"
#define DEFAULTUSERENVREGKEY ".DEFAULT\\Environment"

#define BROADCAST_TIMEOUT (5000)

void warn(char *message) {
	fprintf(stderr, message);
}

int UpdatePathVariable(char *szVarName, char *szPath, int iUser, int iPrepend, int iExpand, int dont_duplicate, int replace)
{
    HKEY hKey = NULL;
    char *szNewPath = NULL;
    char *szExpandedPath = NULL;
	DWORD rv = (DWORD) -1;

	D(("UpdatePathVariable() start"));

	for (;;)
	{
		char szVarValue[MAX_VALUE_LEN];
		DWORD dwPathLen = 0;
		DWORD dwLen = 0;
		DWORD dwType;
		DWORD dwNewLen = 0;

		dwPathLen = strlen(szPath);

		switch (iUser) {
			case 0:
				rv = RegOpenKeyEx(HKEY_LOCAL_MACHINE, ENVREGKEY, 0, KEY_READ | KEY_WRITE | KEY_QUERY_VALUE, &hKey);
				break;
			case 1:
				rv = RegOpenKeyEx(HKEY_CURRENT_USER, USERENVREGKEY, 0, KEY_READ | KEY_WRITE | KEY_QUERY_VALUE, &hKey);
				break;
			case 2:
				rv = RegOpenKeyEx(HKEY_USERS, DEFAULTUSERENVREGKEY, 0, KEY_READ | KEY_WRITE | KEY_QUERY_VALUE, &hKey);
				break;
		}
		if (rv != ERROR_SUCCESS)
		{
			warn(SystemMessage(rv));
			break;
		}

		dwLen = MAX_VALUE_LEN;
		dwType = REG_SZ;
		strcpy(szVarValue, "");
		rv = RegQueryValueEx(hKey, szVarName, NULL, &dwType, (LPBYTE) szVarValue, &dwLen);
		if (rv != ERROR_SUCCESS && rv != ERROR_FILE_NOT_FOUND) {
			dwType = REG_EXPAND_SZ;
			warn(SystemMessage(rv));
			break;
		}

		if (rv == ERROR_FILE_NOT_FOUND) {
			dwLen = 0;
			dwType = iExpand ? REG_EXPAND_SZ : REG_SZ;
		}

		if (dont_duplicate && dwLen) {
			rv = !ERROR_SUCCESS;
			size_t len = strlen(szPath);
			char *p = szVarValue;

			while (p) {
				if (_strnicmp(p, szPath, len) == 0) {
					char *p2 = p + len;
					if (*p2 == ';' || *p2 == '\0') {
						rv = ERROR_SUCCESS;
						break;
					}
				}
				p = strchr(p, ';');
				if (p)
					++p;
			}
			if (rv == ERROR_SUCCESS)
				break;
			rv = ERROR_SUCCESS;
		}

		dwNewLen = dwPathLen + dwLen + 4;
		D(("dwNewLen=%d", dwNewLen));

		szNewPath = (char *) malloc(dwNewLen + 4);
		if (szNewPath == NULL)
		{
			rv = GetLastError();
			warn(SystemMessage(rv));
		    break;
		}

		strcpy(szNewPath, "");

		if (!replace && !iPrepend && dwLen)
		{
			strcpy(szNewPath, szVarValue);
			strcat(szNewPath, ";");
		}

		D(("szNewPath=%s", szNewPath));

		strcat(szNewPath, szPath);

		D(("szNewPath=%s", szNewPath));

		if (!replace && iPrepend && dwLen)
		{
			strcat(szNewPath, ";");
			strcat(szNewPath, szVarValue);
		}

		D(("szNewPath=%s", szNewPath));

		szExpandedPath = (char *) malloc(32767);
		if (szExpandedPath == NULL)
		{
			rv = GetLastError();
			warn(SystemMessage(rv));
			break;
		}

		if (!ExpandEnvironmentStrings(szNewPath, szExpandedPath, 32767)) {
			rv = GetLastError();
			warn(SystemMessage(rv));
			break;
		}

		// todo only update path if running in PE
		if (dwType == REG_EXPAND_SZ)
		{
			rv = RegSetValueEx(hKey, szVarName, 0, dwType, (LPBYTE) szNewPath, strlen(szNewPath) * sizeof(char));
		} else {
			rv = RegSetValueEx(hKey, szVarName, 0, dwType, (LPBYTE) szExpandedPath, strlen(szExpandedPath) * sizeof(char));
		}

		if (rv != ERROR_SUCCESS)
		{
			warn(SystemMessage(rv));
			break;
		}

		if (!SetEnvironmentVariable(szVarName, szExpandedPath))
		{
			rv = GetLastError();
			warn(SystemMessage(rv));
			break;
		}

		static DWORD dummy;
		if (!SendMessageTimeout(HWND_BROADCAST, WM_SETTINGCHANGE, 0, (LPARAM) "Environment", SMTO_ABORTIFHUNG, BROADCAST_TIMEOUT, &dummy))
		{
			rv = GetLastError();
			warn(SystemMessage(rv));
			break;
		}

		rv = ERROR_SUCCESS;
		break;
	}

	if (hKey)
		RegCloseKey(hKey);

	if (szNewPath)
   		free(szNewPath);

	if (szExpandedPath)
   		free(szExpandedPath);

	D(("UpdatePathVariable() returning %d", rv));

    return rv;
}


path_set.cpp
// Copyright (c) 2005-2015 Ross Smith II. See Mit LICENSE in /LICENSE

/*
TODO:

add -d | --delete option, including wildcard:

path_set -d "c:\j2sdk*"

change -d | --default to -f | --default

if the specified path already exists in the path, and they specify --move then it will
	be deleted from its current location and either appended, or prepended
treat repeating \'s as \ in comparing paths
ignore trailing \'s and spaces in comparing paths
change REG type if specified type is different

*/

#ifdef _MSC_VER
#pragma warning(disable:4996)
#pragma comment(lib, "advapi32.lib")
#pragma comment(lib, "user32.lib")
#endif

#ifndef WIN32_LEAN_AND_MEAN
# define WIN32_LEAN_AND_MEAN 1
#endif

#define _WIN32_WINNT 0x501

#include <windows.h>
#include <stdio.h>
#include <cstdlib>
//#include <atlbase.h> // required for Stackwalker

#include "UpdatePathVariable.h"
#include "getopt.h"

#include "version.h"

#define APPNAME			VER_INTERNAL_NAME
#define APPVERSION		VER_STRING2
#define APPCOPYRIGHT	VER_LEGAL_COPYRIGHT

// #define ESCAPE_CHAR '`'
#define ESCAPE_STRING "`"

static char *short_options = "v:sude:nprxzV?";

static struct option long_options[] = {
	{"variable",		required_argument,	0, 'v'},
	{"var",				required_argument,	0, 'v'},
	{"v",				required_argument,	0, 'v'},
	{"system",			no_argument,		0, 's'},
	{"s",				no_argument,		0, 's'},
	{"user",			no_argument,		0, 'u'},
	{"u",				no_argument,		0, 'u'},
	{"default",			no_argument,		0, 'd'},
	{"d",				no_argument,		0, 'd'},
	{"percent",			required_argument,	0, 'e'},
	{"e",				required_argument,	0, 'e'},
	{"no-dups",			no_argument,		0, 'n'},
	{"nodups",			no_argument,		0, 'n'},
	{"no_dups",			no_argument,		0, 'n'},
	{"n",				no_argument,		0, 'n'},
	{"prepend",			no_argument,		0, 'p'},
	{"p",				no_argument,		0, 'p'},
	{"replace",			no_argument,		0, 'r'},
	{"r",				no_argument,		0, 'r'},
	{"reg-expand-sz",	no_argument,		0, 'x'},
	{"reg_expand_sz",	no_argument,		0, 'x'},
	{"expand-sz",		no_argument,		0, 'x'},
	{"expand_sz",		no_argument,		0, 'x'},
	{"expand",			no_argument,		0, 'x'},
	{"x",				no_argument,		0, 'x'},
	{"reg-sz",			no_argument,		0, 'z'},
	{"reg_sz",			no_argument,		0, 'z'},
	{"sz",				no_argument,		0, 'z'},
	{"z",				no_argument,		0, 'z'},
	{"version",			no_argument,		0, 'V'},
	{"V",				no_argument,		0, 'V'},
	{"help",			no_argument,		0, '?'},
	{"h",				no_argument,		0, '?'},
	{"?",				no_argument,		0, '?'},
	{NULL,				0,					0, 0}
};

void usage() {
	printf(
		"Usage: " APPNAME " [options] path [path2 ...]\n\n"
		"Options:\n"
//2345678901234567890123456789012345678901234567890123456789012345678901234567890
"-v | --var var   Append path(s) to environmental variable 'var' (default: Path)\n"
"-s | --system    Modify variable in the system environment (default)\n"
"-u | --user      Modify variable in the current user's environment\n"
"-d | --default   Modify variable in the default user environment\n"
"-e | --percent c Change %% substitution character to 'c' (default is " ESCAPE_STRING ")\n"
"-n | --no-dups   Don't add path if the variable already contains it\n"
"-p | --prepend   Prepend path(s) to variable\n"
"-r | --replace   Replace the value (default is to append)\n"
"-x | --expand-sz Use REG_EXPAND_SZ type (default if " ESCAPE_STRING " char is used)\n"
"-z | --sz        Use REG_SZ type (default)\n"
"-V | --version   Show version and copyright information and quit\n"
"-? | --help      Show this help message and quit\n"

		"\nExamples:\n"
		"  " APPNAME " -n c:\\bin\n"
		"  " APPNAME " -n \"C:\\Program Files\\Java\\bin\"\n"
		"or, even better:\n"
		"  " APPNAME " -n -r -x -v JAVA_HOME " ESCAPE_STRING "ProgramFiles" ESCAPE_STRING "\\Java\n"
		"  " APPNAME " -n " ESCAPE_STRING "JAVA_HOME" ESCAPE_STRING "\\bin\n"
	);
}

int main(int argc, char **argv)
{
#ifdef _DEBUG_ALLOC
    InitAllocCheck();
#endif

	if (argc < 2) {
		usage();
		exit(1);
	}

	int prepend = 0;
	int user = 0;
	int expand = 0;
	int dont_duplicate = 0;
	int replace = 0;

	opterr = 0;

	char *szVarName = "Path";
	char *percent_char = ESCAPE_STRING;

	while (1) {
		int c;
		int option_index = 0;

		if (optind < argc && argv[optind] && argv[optind][0] == '/')
			argv[optind][0] = '-';

		c = getopt_long(argc, argv, short_options, long_options, &option_index);

		if (opterr) {
			usage();
			exit(1);
		}

		if (c == -1)
			break;

        switch (c) {
            case 'V':
				printf("%s - Version %s - %s\n", APPNAME, APPVERSION, __DATE__);
				printf(APPCOPYRIGHT);
				exit(0);
                break;
            case 'd':
				user = 2;
                break;
            case 'e':
				percent_char[0] = (char) optarg[0];
                break;
			case 'n':
				dont_duplicate = 1;
				break;
            case 'p':
				prepend = 1;
                break;
            case 'r':
				replace = 1;
                break;
            case 's':
            	// fallthrough
            case 'm': // undocumented (m=machine)
				user = 0;
                break;
            case 'u':
				user = 1;
                break;
            case 'v':
				szVarName = (char*) optarg;
                break;
            case 'x':
				expand = 1;
                break;
            case 'z':
				expand = 0;
                break;
			case '?':
				usage();
				exit(0);
			case ':':
				fprintf(stderr, "Option -%c requires an operand\n", optopt);
				// fallthrough
			default:
				usage();
				exit(1);
        }
    }

	int rv = 0;
	int found_percent_char = 0;

	for (; optind < argc; ++optind)
	{
		char *arg = argv[optind];
		char *p = arg;
		unsigned int len = strlen(arg);

		while (len-- > 0)
			if (arg[len] == percent_char[0]) {
				found_percent_char = 1;
				arg[len] = '%';
			}

		rv |= UpdatePathVariable(szVarName, arg, user, prepend, expand || found_percent_char, dont_duplicate, replace);
	}

	exit(rv);
	return rv;
}
